### PR TITLE
Mdbenjam add welcome gate

### PR DESCRIPTION
**Pull request by mdbenjam**
Connects: #1999 

Currently users access cases by entering the vacols_id directly into the URL. This PR adds a new page at /reader/appeal that allows the user to choose which appeal they want to work on.

Test Plan:

- [x] Visit reader/appeal and verify it loads the cases assigned to the user.
- [x] Visit reader/appeal and click on one of the cases, make sure you are redirected to the appropriate case.
- [x] Visit reader/appeal and make sure the button text is new/continue based on whether the case has been opened before.

### CODE REVIEW

File: app/controllers/reader/appeal_controller.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        end
+      end
+    end
+  end
+
+  def logo_name
+    "Reader"
+  end
+
+  def appeals
```
Comment (mdbenjam) [inline app/controllers/reader/appeal_controller.rb]: There very well might be a better way to do this. I feel icky about having this in the controller. Open to other ideas.

Comment (mogrenAtWork) [inline app/controllers/reader/appeal_controller.rb]: Yeah, it might be better to break this out into a couple of methods in a service class, `getOpenedAppeals` and `getAllAppeals`, both accept a user as a parameter.

I lack some context, this may be overkill.

Comment (joofsh) [inline app/controllers/reader/appeal_controller.rb]: Could this be moved to the user model? `def assigned_appeals` or something like that. Alternatively we could create a new `Attorney` model, (see the new `Judge` one we have for hearings)

File: app/controllers/reader/documents_controller.rb [hunk 2]
```diff
@@ -1,10 +1,19 @@
-class Reader::DocumentsController < ApplicationController
+class Reader::DocumentsController < Reader::AppealController
```
Comment (joofsh) [inline app/controllers/reader/documents_controller.rb]: Why does this inherit from `Reader::AppealController`?  Ideally the "reader application controller" would only have methods that are shared across both sub-controllers. If we're trying to share the `logo_path` and `logo_name` method, what about creating a `Reader::ApplicationController`? That could be a convention going forward for all apps

File: db/migrate/20170615162530_create_appeal_view.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+class CreateAppealView < ActiveRecord::Migration
+  # This is a new table, so the new indecies will be fine
+  safety_assured
+  
+  def change
+    create_table :appeal_views do |t|
+      t.belongs_to :user, null: false
+      t.belongs_to :appeal, null: false
+
+      t.datetime   :first_viewed_at
```
Comment (joofsh) [inline db/migrate/20170615162530_create_appeal_view.rb]: Two things:
- We should be adding timestamps for all tables going forward (created_at & updated_at)
- Based on this table's usage, this appears to be the same as the records `created_at`. I think if we add base timestamps it'll simplify some of the controller code

File: spec/models/appeal_spec.rb [hunk 4]
```diff
@@ -902,4 +902,25 @@
...
+      Appeal.create_appeal_without_lazy_load(vacols_id: saved_appeal.vacols_id,
+                                             signed_date: date)
+    end
+
+    it "creates an appeals object with attributes" do
+      expect(appeal.signed_date).to eq(date)
+    end
+
+    it "appeal does not lazy load vacols data" do
+      expect(appeal.veteran_first_name).to be_nil
```
Comment (joofsh) [inline spec/models/appeal_spec.rb]: This doesn't really test that we didn't lazy load data from VACOLS (the veteran's name could be nil in VACOLS). Can we instead test more directly that `check_and_load_vacols_data!` is not called?

File: app/models/appeal.rb [hunk 5]
```diff
@@ -424,6 +428,13 @@ def convert_file_number_to_vacols(file_number)
       fail Caseflow::Error::InvalidFileNumber
     end
 
+    def create_appeal_without_lazy_load(hash)
```
Comment (joofsh) [inline app/models/appeal.rb]: - Can we rename this from `create_` to something else (since we aren't actually persisting to the DB). Maybe just `initialize_`
- After looking through this PR more, do we actually need to disable lazy loading? Presumably in this request cycle, you're not calling any other appeal attributes that would come back `nil`. If we are turning off the lazy loading, then we should give the developer some kind of warning that they tried to access a value that we never fetched from VACOLS, perhaps even throwing an error?

File: client/app/reader/DecisionReviewer.jsx [hunk 6]
```diff
@@ -144,11 +145,15 @@ export class DecisionReviewer extends React.PureComponent {
     ;
   }
 
+  routedCaseSelect = () => {
```
Comment (joofsh) [inline client/app/reader/DecisionReviewer.jsx]: Channeling my innner- @NickHeiner , this could be a 1-liner

File: client/app/components/Link.jsx [hunk 7]
```diff
@@ -0,0 +1,53 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Link as ReduxLink } from 'react-router-dom';
+
+export default class Link extends React.Component {
+  className = (button) => {
+    switch (button) {
```
Comment (joofsh) [inline client/app/components/Link.jsx]: This looks like it could be an object mapping:
```
MAPPING = {
  primary: 'usa-button',
  secondary: 'usa-button-outline',
  ...
}
MAPPING[button] || ''
```

File: client/app/components/Link.jsx [hunk 8]
```diff
@@ -0,0 +1,53 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Link as ReduxLink } from 'react-router-dom';
```
Comment (joofsh) [inline client/app/components/Link.jsx]: nit, but this is probably misnamed. This is a react router link, totally disconnected from redux

File: app/controllers/reader/appeal_controller.rb [hunk 9]
```diff
@@ -0,0 +1,40 @@
+class Reader::AppealController < ApplicationController
+  def index
+    respond_to do |format|
+      format.html { return render(:index) }
```
Comment (joofsh) [inline app/controllers/reader/appeal_controller.rb]: nit: do we need the `return` here? We don't have one in the json block (or most other render calls). This should naturally be the last line of execution

File: spec/feature/reader_spec.rb [hunk 10]
```diff
@@ -104,6 +104,33 @@ def add_comment(text)
...
+
+        click_on "New", match: :first
+
+        expect(page).to have_content("Documents")
+
+        click_on "Caseflow Reader"
+
+        click_on "Continue"
+
+        expect(page).to have_content("Documents")
```
Comment (joofsh) [inline spec/feature/reader_spec.rb]: Is this a new page? Can we test the url change?

File: lib/fakes/case_assignment_repository.rb [hunk 11]
```diff
@@ -1,37 +1,33 @@
...
-        signed_date: nil
-      },
-      {
-        vacols_id: "2725685",
-        date_assigned: "2013-05-31 00:00:00 UTC".to_datetime,
-        date_received: nil,
-        signed_date: nil
-      }
+    appeal_records || [
+      Appeal.create_appeal_without_lazy_load(vacols_id: "reader_id1",
```
Comment (joofsh) [inline lib/fakes/case_assignment_repository.rb]: can we move this out to a separate method for clarity? `appeal_records || default_appeal_records` or something

File: spec/models/concerns/associated_vacols_model_spec.rb [hunk 12]
```diff
@@ -60,6 +60,15 @@ def self.repository
     end
   end
 
+  context "#turn_off_lazy_loading" do
+    before { model.turn_off_lazy_loading }
+
+    it "will not load data when check_and_load_vacols_data is called" do
+      expect(TestVacolsModelRepository).to receive(:load_vacols_data).exactly(0).times
```
Comment (joofsh) [inline spec/models/concerns/associated_vacols_model_spec.rb]: 👍 nice

File: client/app/reader/DecisionReviewer.jsx [hunk 13]
```diff
@@ -159,7 +164,10 @@ export class DecisionReviewer extends React.PureComponent {
     const Router = this.props.router || BrowserRouter;
 
     return <Router basename="/reader/appeal" {...this.props.routerTestProps}>
-      <Route path="/:vacolsId/documents" render={this.rootRoute} />
+      <div className="section--document-list">
+        <Route path="/:vacolsId/documents" render={this.rootRoute} />
```
Comment (joofsh) [inline client/app/reader/DecisionReviewer.jsx]: I don't remember where the method name came from, possibly me! But `rootRoute` seems like a bad name now, given that the `routedCaseSelect` is the true rootRoute here

File: client/app/components/Link.jsx [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    let {
+      to,
+      button,
+      children
+    } = this.props;
+
+    const type = button ? 'button' : null;
+    const className = this.className(button);
+
+    if (button === 'disabled') {
```
Comment (mogrenAtWork) [inline client/app/components/Link.jsx]: I can't find where this is used, but I may be overlooking something

Comment (mogrenAtWork) [inline client/app/components/Link.jsx]: I can't find where a Link is disabled, but I may be overlooking something. Is the use case for this that the URL for the link is currently being viewed?

File: client/test/karma/reader/reducer-test.js [hunk 15]
```diff
@@ -285,10 +285,13 @@ describe('Reader reducer', () => {
       const state = reduceActions([
         {
           type: Constants.RECEIVE_DOCUMENTS,
-          payload: [{
-            id: docId,
-            tags: []
-          }]
+          payload: {
+            documents: [{
```
Comment (mogrenAtWork) [inline client/test/karma/reader/reducer-test.js]: Do you have any tests that test when `documents` is null or empty?

File: app/controllers/reader/documents_controller.rb (hunk 2 above)
Comment (mdbenjam) [inline app/controllers/reader/documents_controller.rb]: I like that a lot.

File: client/app/components/Link.jsx (hunk 8 above)
Comment (mdbenjam) [inline client/app/components/Link.jsx]: Gah! You're right. Nice catch.

File: client/app/components/Link.jsx (hunk 14 above)
Comment (mdbenjam) [inline client/app/components/Link.jsx]: It is not. I just know that we often use disabled buttons so I thought I'd add the styling here. Some context--we have a button component that we (1) use when we should really use a link (this is a result of our introduction of react router after building a hacked together SPA). (2) The styling on button is way to free form. I want it all built in. I'm hoping this styling will get us part of the way there for links.

File: client/app/reader/DecisionReviewer.jsx (hunk 6 above)
Comment (mdbenjam) [inline client/app/reader/DecisionReviewer.jsx]: You do a pretty good Nick.

File: spec/models/appeal_spec.rb (hunk 4 above)
Comment (mdbenjam) [inline spec/models/appeal_spec.rb]: I actually disagree with this. The record we're using explicitly does have a first name. "George". And it feels like checking check_and_load_vacols_data is not called is too deep in the implementation.

File: client/app/components/Link.jsx (hunk 14 above)
Comment (mdbenjam) [inline client/app/components/Link.jsx]: If you think this is outside of the scope of this PR. We can remove it.

File: spec/models/appeal_spec.rb (hunk 4 above)
Comment (joofsh) [inline spec/models/appeal_spec.rb]: okay I see what you're saying. It's pretty explicit in the test setup 👍

File: client/test/karma/reader/reducer-test.js (hunk 15 above)
Comment (mdbenjam) [inline client/test/karma/reader/reducer-test.js]: Added some above.

File: app/models/concerns/associated_vacols_model.rb [hunk 16]
```diff
@@ -1,6 +1,8 @@
 module AssociatedVacolsModel
   extend ActiveSupport::Concern
 
+  class LazyLoadingTurnedOff < StandardError; end
```
Comment (joofsh) [inline app/models/concerns/associated_vacols_model.rb]: Minor nit, but it's (imo) idiomatic ruby to end error classes as `Error`. I think it adds clarity but we def don't do it consistently in Caseflow, thoughts?

```
app/models/vacols/note.rb
5:  class InvalidNoteCodeError < StandardError; end
6:  class InvalidNotelengthError < StandardError; end
7:  class TextRequiredError < StandardError; end
```

File: app/models/concerns/associated_vacols_model.rb [hunk 17]
```diff
@@ -42,8 +50,9 @@ def vacols_record_exists?
     check_and_load_vacols_data!
   end
 
-  def turn_off_lazy_loading
+  def turn_off_lazy_loading(initial_values: nil)
     @vacols_load_status = :disabled
+    @initial_values = initial_values
```
Comment (joofsh) [inline app/models/concerns/associated_vacols_model.rb]: What's the value in using `@initial_values` versus the existing `attr_accessor`s? This seems to add complexity by offering a new path/mechanism for storing the data. I think there's value in consistently storing it on `@veteran_first_name` whether it was lazy loaded or not

File: db/migrate/20170615162530_create_appeal_view.rb [hunk 18]
```diff
@@ -7,7 +7,8 @@ def change
       t.belongs_to :user, null: false
       t.belongs_to :appeal, null: false
 
-      t.datetime   :first_viewed_at
+      t.datetime   :created_at
```
Comment (joofsh) [inline db/migrate/20170615162530_create_appeal_view.rb]: Hmm, does this work? I'm not sure if Rails will know to automatically set these columns this way. I think you need to use `t.timestamps`. For consistency, it would also be good to set `null: false` on them.  https://github.com/department-of-veterans-affairs/caseflow/blob/8033440ecdf88cf05bb43393616856452ee21bf2/db/migrate/20170414183652_create_tags.rb#L7

File: spec/models/user_spec.rb [hunk 19]
```diff
@@ -190,6 +190,40 @@
     end
   end
 
+  context "#current_case_assignments_with_views" do
```
Comment (joofsh) [inline spec/models/user_spec.rb]: Nice! this feels way better tested 👍

File: app/models/concerns/associated_vacols_model.rb (hunk 16 above)
Comment (mdbenjam) [inline app/models/concerns/associated_vacols_model.rb]: That makes sense to me.

File: app/models/concerns/associated_vacols_model.rb (hunk 17 above)
Comment (mdbenjam) [inline app/models/concerns/associated_vacols_model.rb]: I agree--but it was hard to get working. How do we know which ones were set? We need that information to be able to through errors if you try to load from vacols.

File: client/app/reader/CaseSelect.jsx [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+      }
+    ];
+  }
+
+  getKeyForRow = (index, row) => {
+    return row.vacols_id;
+  }
+
+  render() {
+    if (!this.props.assignments) {
```
Comment (joofsh) [inline client/app/reader/CaseSelect.jsx]: Something to consider for future PRs. We may want to consider moving away from this "if bad props don't render" style. This is an indicator we need a higher order "connected component". That higher level component would contain the `componentDidMount` logic, and determine whether to internally render the dumb (UI) component with the proper props. That way `assignments` can be a required prop

File: app/models/concerns/associated_vacols_model.rb [hunk 21]
```diff
@@ -42,6 +45,14 @@ def vacols_record_exists?
     check_and_load_vacols_data!
   end
 
+  def turn_off_lazy_loading(initial_values: nil)
+    @vacols_load_status = :disabled
+    @initial_values = initial_values
```
Comment (joofsh) [inline app/models/concerns/associated_vacols_model.rb]: small nit. Can we save only the keys here, and name it something more descriptive

File: client/app/components/Link.jsx (hunk 14 above)
Comment (mogrenAtWork) [inline client/app/components/Link.jsx]: No, I think it's great! Just wanted to make sure it was added intentionally and wasn't left hanging around.

File: client/app/components/Link.jsx [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+
+const CLASS_NAME_MAPPING = {
+  primary: 'usa-button',
+  secondary: 'usa-button-outline',
+  disabled: 'usa-button-disabled'
+};
+
+export default class Link extends React.Component {
+  render() {
+    let {
```
Comment (NickHeiner) [inline client/app/components/Link.jsx]: This should be `const`.

File: client/app/reader/CaseSelect.jsx [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+            to={`/${row.vacols_id}/documents`}>
+              {buttonText}
+            </Link>;
+        }
+      }
+    ];
+  }
+
+  getKeyForRow = (index, row) => {
+    return row.vacols_id;
```
Comment (NickHeiner) [inline client/app/reader/CaseSelect.jsx]: Please use the short form instead of the long form for single-expression functions.

File: client/app/reader/CaseSelect.jsx [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+  }
+
+  render() {
+    if (!this.props.assignments) {
+      return <div></div>;
```
Comment (NickHeiner) [inline client/app/reader/CaseSelect.jsx]: `return null` instead of an empty `div`.

File: client/app/reader/CaseSelect.jsx [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+    ApiUtil.get('/reader/appeal').then((response) => {
+      const returnedObject = JSON.parse(response.text);
+
+      this.props.onReceiveAssignments(returnedObject);
+    }, this.props.onInitialDataLoadingFail);
+  }
+}
+
+const mapStateToProps = (state) => {
+  return _.pick(state, 'assignments');
```
Comment (NickHeiner) [inline client/app/reader/CaseSelect.jsx]: Please use the short form.

File: client/app/reader/DecisionReviewer.jsx [hunk 26]
```diff
@@ -144,11 +145,13 @@ export class DecisionReviewer extends React.PureComponent {
...
 
-  rootRoute = (props) => {
+  routedCaseSelect = () => <CaseSelect />
+
+  documentsRoute = (props) => {
     const { vacolsId } = props.match.params;
 
     return <LoadingScreen vacolsId={vacolsId}>
-      <div className="section--document-list">
+      <div>
```
Comment (NickHeiner) [inline client/app/reader/DecisionReviewer.jsx]: Is this `div` necessary if it has no `class` or `id`?

File: client/test/karma/reader/reducer-test.js [hunk 27]
```diff
@@ -1,13 +1,56 @@
+/* eslint-disable max-lines */
```
Comment (NickHeiner) [inline client/test/karma/reader/reducer-test.js]: I would also be happy to create a directory called `test/karma/reader/reducer/` and have files like `receive-documents.js`, so each action was tested separately, and we didn't need to have a massive file. But that can happen in a later PR.

File: client/test/karma/reader/reducer-test.js [hunk 28]
```diff
@@ -1,13 +1,56 @@
...
 import { expect } from 'chai';
 import { reducer } from '../../../app/reader/reducer';
 import * as Constants from '../../../app/reader/constants';
 
 /* eslint-disable no-undefined */
+/* eslint-disable no-unused-expressions */
```
Comment (NickHeiner) [inline client/test/karma/reader/reducer-test.js]: I would prefer not to disable this lint rule. I don't like the Chai form that involves "unused" expressions, because if you have a typo, the test will pass anyway:

```js
// totally passes
expect(foo).to.be.undeifned
```

I recommend against the use of this syntax. We should just do `expect(foo).to.equal(undefined)` instead.

File: db/migrate/20170615162530_create_appeal_view.rb [hunk 29]
```diff
@@ -0,0 +1,16 @@
+class CreateAppealView < ActiveRecord::Migration
+  # This is a new table, so the new indecies will be fine
```
Comment (NickHeiner) [inline db/migrate/20170615162530_create_appeal_view.rb]: "indecies" is spelled wrong.

File: client/test/karma/reader/reducer-test.js (hunk 27 above)
Comment (mdbenjam) [inline client/test/karma/reader/reducer-test.js]: Agreed

File: client/app/reader/DecisionReviewer.jsx (hunk 26 above)
Comment (mdbenjam) [inline client/app/reader/DecisionReviewer.jsx]: Yes it is. You can only pass one component in as children.
